geowebcache/core/src/main/java/org/geowebcache/storage/blobstore/file/FileBlobStore.java

```diff
@@ -644,27 +661,104 @@ private static String urlDecUtf8(String value) {
+
+        // Evaluate updating key arg related values (remove/add/update or skip)
+        boolean skipUpdate = false;
+        // Note: since we want to avoid writing metadata file if nothing changed, then we need to
+        // check if it's required
+        // or not (flag skipUpdate). That is reason not to use metadata.merge method instead of
+        // using setProperty
         if (null == value) {
             metadata.remove(key);
         } else {
+            String currValue = metadata.getProperty(key);
+            String encodedValue = null;
             try {
-                metadata.setProperty(key, URLEncoder.encode(value, "UTF-8"));
+                encodedValue = URLEncoder.encode(value, "UTF-8");
             } catch (UnsupportedEncodingException e) {
                 throw new RuntimeException(e);
             }
+            skipUpdate =
+                    currValue != null && encodedValue != null && currValue.equals(encodedValue);
+            // If existing value for key is the same as value arg, then avoid to write it
+            if (!skipUpdate) {
+                metadata.setProperty(key, encodedValue);
+            } else {
+                skipUpdate = true;
+            }
         }
 
-        final File metadataFile = getMetadataFile(layerName);
+        // We call to save metadata only if has changed
+        if (!skipUpdate) {
+            // Write metadata in a temp file
+            File tempFile = null;
+            try {
+                tempFile = writeTempMetadataFile(metadata);
+            } catch (IOException e) {
+                log.error("Cannot create temporary file for " + metadataFile.getPath());
+                throw new UncheckedIOException(e);
+            }
+            // Attempt to rename tmp file to metadata file
+            int tryagain = 0;
+            while (tryagain < FileBlobStore.METADATA_MAX_RW_ATTEMPTS) {
+                // Read current metadatafile date modification (currentModTime)
+                long currentModDate = metadataFile.lastModified();
+                if (lastModDate == 0 || lastModDate == currentModDate) {
+                    if (!metadataFile.getParentFile().exists()) {
+                        metadataFile.getParentFile().mkdirs();
+                    }
+                    // atomic rename of temporary metadata file with getMetadataFilename
+                    // file, in such case we'll just eliminate this one
+                    if (FileUtils.renameFile(tempFile, metadataFile)) {
+                        tempFile = null;
+                    } else {
+                        // Renaming failed !
+                        log.error("Error while renaming metadata file " + metadataFile.getPath());
+                    }
+                    tryagain = FileBlobStore.METADATA_MAX_RW_ATTEMPTS;
+                } else {
+                    // Try again since some other GWC updated the file in middle of writing process
+                    ++tryagain;
```

Comment:
Not wrong, but a bit out of place, the idiomatic way to perform a stand alone counter increment is "tryagain++"